e2e/tests/test_snapp_deposit_withdraw.rs

```diff
@@ -0,0 +1,152 @@
+use ethcontract::web3::api::Web3;
+use ethcontract::web3::transports::Http;
+use ethcontract::web3::types::{H160, H256, U128, U256};
+use ethcontract::Account;
+use std::str::FromStr;
+
+#[test]
+fn test_deposit_and_withdraw() {
+    let (eloop, http) = Http::new("http://localhost:8545").expect("transport failed");
+    eloop.into_remote();
+    let web3 = Web3::new(http);
+    let (instance, accounts, tokens, db) = setup_snapp(&web3, 3, 3);
+
+    // Test environment values
+    let deposit_amount = 18_000_000_000_000_000_000u128;
+
+    let user_address = &accounts[2];
+    let user_id = instance
+        .public_key_to_account_map(*user_address)
+        .call()
+        .wait()
+        .expect("Could not recover account id");
+    let db_account_id = H160::from_low_u64_be(user_id);
+    // read_balance expects u16, while ethcontracts only accepts u64.
+    let token_id = 2u16;
+    let initial_balance = tokens[token_id as usize]
+        .balance_of(*user_address)
+        .call()
+        .wait()
+        .expect("Could not retrieve token balance");
+    println!("Initial Balance: {:?}", initial_balance);
+
+    let initial_state_hash = instance
+        .get_current_state_root()
+        .call()
+        .wait()
+        .expect("Could not recover initial state hash");
+
+    instance
+        .deposit(token_id.into(), U128::from(deposit_amount))
+        .from(Account::Local(*user_address, None))
+        .send()
+        .wait()
+        .expect("Failed to send first deposit");
+
+    wait_for(&web3, 181);
+
+    // Check that contract was updated
+    println!("Waiting for deposit state transition");
+    let expected_deposit_hash =
+        H256::from_str("781cff80f5808a37f4c9009218c46af3d90920f82110129f6d925fafb3b23f2d").unwrap();
+
+    let after_deposit_state = await_state_transition(&instance, &initial_state_hash);
+    assert_eq!(
+        expected_deposit_hash,
+        H256::from_slice(&after_deposit_state)
+    );
+
+    // Check that DB was updated (with correct balances)
+    wait_for_condition(|| {
+        db.get_balances_for_state_root(&expected_deposit_hash)
+            .is_ok()
+    })
+    .expect("Deposit: Did not detect expected DB update");
+    let state = db
+        .get_balances_for_state_root(&expected_deposit_hash)
+        .unwrap();
+
+    // TODO - Our storage for AccountState should use account_id properly and NOT H160!
+    // The account id here is counter intuitive. (since we have to increment by 1)
+    assert_eq!(state.read_balance(token_id, db_account_id), deposit_amount);
+
+    instance
+        .request_withdrawal(token_id.into(), U128::from(deposit_amount))
+        .from(Account::Local(*user_address, None))
+        .send()
+        .wait()
+        .expect("Failed to request withdraw");
+
+    wait_for(&web3, 181);
+    let expected_withdraw_hash =
+        H256::from_str("7b738197bfe79b6d394499b0cac0186cdc2f65ae2239f2e9e3c698709c80cb67").unwrap();
```

Comment:
Should this maybe be calculated somehow eventually?

Reply:
It is calculated by the driver. We are expecting this to be the state has and this is (thus) meant to be a hardcoded value.